lib/Transforms/ConvertSecretWhileToStaticFor/ConvertSecretWhileToStaticFor.cpp

```diff
@@ -0,0 +1,161 @@
+#include "mlir/include/mlir/Support/LLVM.h"           // from @llvm-project
+#include "mlir/include/mlir/Support/LogicalResult.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_CONVERTSECRETWHILETOSTATICFOR
+#include "lib/Transforms/ConvertSecretWhileToStaticFor/ConvertSecretWhileToStaticFor.h.inc"
+
+struct SecretWhileToStaticForConversion : OpRewritePattern<scf::WhileOp> {
+  using OpRewritePattern<scf::WhileOp>::OpRewritePattern;
+
+ public:
+  SecretWhileToStaticForConversion(DataFlowSolver *solver, MLIRContext *context)
+      : OpRewritePattern(context), solver(solver){};
+
+  // TODO(#846): Add support for do-while loops
+  LogicalResult matchAndRewrite(scf::WhileOp whileOp,
+                                PatternRewriter &rewriter) const override {
+    auto whileCondition = whileOp.getConditionOp()->getOperand(0);
+
+    auto *conditionSecretnessLattice =
+        solver->lookupState<SecretnessLattice>(whileCondition);
+
+    if (!conditionSecretnessLattice) {
+      InFlightDiagnostic diag = mlir::emitWarning(
+          whileOp.getLoc(),
+          "Secretness for scf.while condition has not been set:");
+      return failure();
+    }
+
+    bool isConditionSecret =
+        conditionSecretnessLattice->getValue().getSecretness();
+
+    // If condition is not secret, no transformation is needed
+    if (!isConditionSecret) {
+      return failure();
+    }
+
+    auto maxIterAttr = whileOp->getAttrOfType<IntegerAttr>("max_iter");
+
+    // If maximum iteration attribute is not provided, emit a warning and return
+    // failure
+    if (!maxIterAttr) {
+      InFlightDiagnostic diag = mlir::emitWarning(
+          whileOp.getLoc(),
+          "Cannot convert secret scf.while to static affine.for "
+          "since a static maximum iteration attribute has not been provided:");
+
+      return failure();
+    }
+
+    ImplicitLocOpBuilder builder(whileOp->getLoc(), rewriter);
+
+    int upperBound = maxIterAttr.getInt();
+
+    SmallVector<Value> iterArgs(whileOp.getInits());
+
+    auto newForOp =
+        builder.create<affine::AffineForOp>(0, upperBound, 1, iterArgs);
+    newForOp->setAttrs(whileOp->getAttrs());
+
+    builder.setInsertionPointToStart(newForOp.getBody());
+
+    IRMapping mp;
+    for (BlockArgument blockArg : whileOp.getBody()->getArguments()) {
+      mp.map(blockArg,
+             newForOp.getBody()->getArguments()[blockArg.getArgNumber() + 1]);
+    }
+
+    Value condition;
+    for (auto &op : whileOp.getBefore().getOps()) {
+      if (!isa<scf::ConditionOp>(op)) {
+        // Copy op in "before" region to the forOp's body
+        builder.clone(op, mp);
+      } else {
+        // Get condition used in scf.condition op
+        condition = mp.lookup(dyn_cast<scf::ConditionOp>(op).getOperand(0));
+      }
```

Comment:
nit: slight simplification. dyn_cast returns nullptr if the cast fails.

```suggestion
      if (auto conditionOp = dyn_cast<scf::ConditionOp>(op)) {
        // Get condition used in scf.condition op
        condition = mp.lookup(conditionOp.getOperand(0));       
      } else {
        // Copy op in "before" region to the forOp's body
        builder.clone(op, mp);
      }
```